### ISSUE TITLE
"Start index cannot be less than zero" error while sending Unicode subject via SMTP

### ISSUE THREAD

**Issue by dukesaeed** [bertjohnson/OpaqueMail#44]
Hi,
both your test platform and my code return this error while sending Smtp with a Unicode subject,

the subject sample that causes error:

یک بروزرسانی در نرم افزار سامانه پایش آب ایران (IWMS) توسط سعید ایجاد شد.

my code:

```
SmtpClient smtpClient = new SmtpClient(ServerSettings.SMTPHost, 587);
            smtpClient.Credentials = new  NetworkCredential(ServerSettings.SMTPUsername, ServerSettings.SMTPPassword);
            smtpClient.EnableSsl = true;

            MailMessage message = new MailMessage(item.FromAddress, item.ToAddress, item.MailSubject, item.Body);
            message.IsBodyHtml = true;
            message.SubjectEncoding = Encoding.Unicode;

            try
            {
                await smtpClient.SendAsync(message);
            }
            catch
            {
                return false;
            }

```

**Comment by bertjohnson** [comment]
Hi dukesaeed, thanks for the report. I tried to recreate the issue by sending emails with that subject, which encoded as follows:

`=?UTF-8?B?24zaqSDYqNix2YjYstix2LPYp9mG24wg2K/YsSDZhtix2YUg2KfZgdiy2KfYsSDYs9in?=
	=?UTF-8?B?2YXYp9mG2Ycg2b7Yp9uM2LQg2KLYqCDYp9uM2LHYp9mGIChJV01TKSDYqtmI2LPYtyDYs9i524zYryA=?=
	=?UTF-8?B?2KfbjNis2KfYryDYtNivLg==?=`

It looks like OpaqueMail can parse that properly, but I'm guessing the email you received had a different encoding.

If you still have that email handy, could you please view the email's source and share the encoded version of the subject (which will start with `=?` and end with `?=`)? If you see the line number of the error, that will help too. Thanks for your help debugging.

**Comment by bertjohnson** [comment]
Without an example, we won't be able to debug this. Closing this issue, but feel free to re-open with any examples. Thanks.

**Comment by sabellefallEFC** [comment]
Hi,

I resurrected this topic because I'm dealing with a similar problem.
In some cases (randomly?), the "cr lf" characters are in the middle of a word. In that case, the word isn't decoded correctly, it seems.

The subject was :
![image](https://cloud.githubusercontent.com/assets/11231456/21710709/b44f8264-d3ea-11e6-9544-549c5c29e86e.png)

-> can't send a message by SMTP with that subject, getting the "Start index cannot be less than zero" exception message.

Here's another example with the message body.

The original e-mail HTML source :
![image](https://cloud.githubusercontent.com/assets/11231456/21710411/b67a9cec-d3e8-11e6-8da1-cc30e47aabc8.png)

This is the message.Body that I get :
![image](https://cloud.githubusercontent.com/assets/11231456/21710313/190ed950-d3e8-11e6-8daf-71bf0eb6468c.png)

I noticed I had trouble with my Regex matching to get the field values. There were also sometimes carriage returns in the middle of these values, so the matching failed.

Like that example (original mail in plain text)
![image](https://cloud.githubusercontent.com/assets/11231456/21710522/88c8d916-d3e9-11e6-8929-4216c6514021.png)

-> Can't get it my matching from "Prénom : " to "/r/n"

So far I'm using workarounds, and not using the subjects as is, but any help would be appreciated.

**Comment by bertjohnson** [comment]
Hi sabellefallEFC, thanks for the report. Would you be able to forward me one of the problematic messages? You can send to bert at opaquemail dot org.

For headers. there's the concept of "folding" metadata over the course of multiple lines. A long subject can be wrapped onto a second line as long as the next line starts with a space or tab. It sounds like that's not being honored for some reason.

I can look into the body encoding as well with an example. Thanks.

**Comment by sabellefallEFC** [comment]
Hello,

 

Thank you for your quick reply !

You’ll find attached 2 examples of these problematic e-mails.

 

Best regards,

 


Description : Description : signature

Sabelle FALL 
Référente RED / .NET

Service Informatique

 

 

De : Bert Johnson [mailto:notifications@github.com] 
Envoyé : vendredi 6 janvier 2017 21:47
À : bertjohnson/OpaqueMail
Cc : FunkyDev; Comment
Objet : Re: [bertjohnson/OpaqueMail] "Start index cannot be less than zero" error while sending Unicode subject via SMTP (#44)

 

Hi sabellefallEFC, thanks for the report. Would you be able to forward me one of the problematic messages? You can send to bert at opaquemail dot org.

For headers. there's the concept of "folding" metadata over the course of multiple lines. A long subject can be wrapped onto a second line as long as the next line starts with a space or tab. It sounds like that's not being honored for some reason.

I can look into the body encoding as well with an example. Thanks.

**Comment by bertjohnson** [comment]
Thanks for following up via email, sabellefallEFC. WIth your help, I found an error when wrapping headers >78 bytes that contained high-byte characters. That's fixed in release 2.4.3.

**Comment by sabellefallEFC** [comment]
Thank you for the quick fix ! Very much appreciated.
